packages/block-editor/src/components/inserter/search-results.js

```diff
@@ -70,6 +71,21 @@ function InserterSearchResults( {
 		destinationRootClientId
 	);
 
+	const filteredBlockPatterns = useMemo( () => {
+		const results = searchItems( patterns, filterValue );
+		return maxBlockPatterns !== undefined
+			? results.slice( 0, maxBlockPatterns )
+			: results;
+	}, [ filterValue, patterns, maxBlockPatterns ] );
+
+	let maxBlockTypesToShow = maxBlockTypes;
+	if (
+		__experimentalPrioritizePatterns &&
+		filteredBlockPatterns.length > 2
+	) {
+		maxBlockTypesToShow = 0;
```

Comment:
Nit: For both filteredPatterns and filteredBlocks, we could bail early if the max is 0 (for perf reasons but really not sure whether it has any impact)